server/text_generation_server/models/ct2_causal_lm.py

```diff
@@ -0,0 +1,394 @@
+# coding=utf-8
+# Copyright 2023 HuggingFace Inc. team and Michael Feil. All rights reserved.
+#
+# This code is based on Text-generation-inference causal_lm.py implementation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+import numpy as np
+import os
+import multiprocessing
+from pathlib import Path
+
+from huggingface_hub.constants import HUGGINGFACE_HUB_CACHE
+from opentelemetry import trace
+from transformers import (
+    AutoTokenizer,
+    AutoConfig,
+)
+from typing import Optional, Tuple, List, Type, Dict
+
+from text_generation_server.models import Model
+from text_generation_server.models.types import (
+    PrefillTokens,
+    Generation,
+    GeneratedText,
+)
+
+from text_generation_server.utils import Sampling
+from text_generation_server.models.causal_lm import CausalLMBatch
+
+try:
+    import ctranslate2
+except ImportError:
+    ctranslate2 = None
+
+
+tracer = trace.get_tracer(__name__)
+
+
+class CT2CausalLM(Model):
+    def __init__(
+        self,
+        model_id: str,
+        revision: Optional[str] = None,
+        quantize: Optional[str] = None,
+        dtype: Optional[torch.dtype] = None,
+        trust_remote_code: bool = False,
+    ):
+        if ctranslate2 is None:
+            raise ValueError(
+                "for quantization with ct2, the installation requires the pip package ctranslate2. "
+                "install via `text-generation-server[ct2]` or pip install ctranslate2 is required.",
+            )
+
+        tokenizer = AutoTokenizer.from_pretrained(
+            model_id,
+            revision=revision,
+            padding_side="left",
+            truncation_side="left",
+            trust_remote_code=trust_remote_code,
+        )
+
+        # Start CT2
+        ct2_generator_kwargs = {"inter_threads": 1}
+        if torch.cuda.is_available():
+            self.ct2_device = "cuda"
+        else:
+            self.ct2_device = "cpu"
+            ct2_generator_kwargs["intra_threads"] = multiprocessing.cpu_count() // 2
```

Comment:
Would be great if this could be also set by an env var.

Reply:
I would suggest to default to this value (default is 0, whereas taking the physical cores has ~50-70% faster inference speed)

However I like the idea of being able to set this an env variable.

Reply:
Updated now, if you are happy with the solution, can you resolve this comment.

Reply:
LGTM, though you'll probably need to convert the string obtained from os.environ into an int!